### PR TITLE
Add upgrade routine to remove old action schedule token

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
Add upgrade routine to remove old action schedule token - which is no longer of use post https://github.com/civicrm/civicrm-core/pull/20961 which ensures standardised alternatives exist

This adds a routing to swap out the action schedule tokens on upgrade.

Given the token is completely invisible in the UI it's unlikely it is in use anywhere but
rather than just remove it from Contribute_Tokens this adds the upgrade
routine to remove this token, and others

Before
----------------------------------------
`{contribution.status} ` token is part of Contribution token processor - but not exposed in the UI & this token processor does not 'listen to anything other than scheduled reminders so we can be comfortable that if we remove it from scheduled reminders in the DB we can also remove from the class

After
----------------------------------------
poof

Technical Details
----------------------------------------
Note that in the similar upgrade for message templates I added an message for when the string wasn't cleaned up but that applied to smarty but not to 'token tokens' - ie we can't be sure that swapping out `{$displayName}` gets everything because `{if $displayName...` is valid but `{contribution.status}` will always be exactly that

Comments
----------------------------------------

### CODE REVIEW

File: CRM/Upgrade/Incremental/php/FiveFortyOne.php [hunk 1]
```diff
@@ -65,6 +65,12 @@ public function setPostUpgradeMessage(&$postUpgradeMessage, $rev) {
   public function upgrade_5_41_alpha1($rev) {
     $this->addTask(ts('Upgrade DB to %1: SQL', [1 => $rev]), 'runSql', $rev);
     $this->addTask('Install legacy custom search extension', 'installCustomSearches');
+    $this->addTask('Replace legacy displayName smarty token in Invoice workflow template',
```
Comment (eileenmcnaughton) [inline CRM/Upgrade/Incremental/php/FiveFortyOne.php:68]: This line is not part of this PR but it was merged & then lost when the line above was merged so this puts it back

File: tests/phpunit/CRM/Contribute/ActionMapping/ByTypeTest.php [hunk 2]
```diff
@@ -260,7 +260,6 @@ public function testTokenRendering(): void {
       first name = {contact.first_name}
       receive_date = {contribution.receive_date}
       contribution status id = {contribution.contribution_status_id}
-      legacy style status = {contribution.status}
```
Comment (seamuslee001) [inline tests/phpunit/CRM/Contribute/ActionMapping/ByTypeTest.php:263]: @eileenmcnaughton I know you put this in here but is there anything that points conclusively that we are meant to be using name not label from this context (just figuring that the default install will have label = name for completed)

Comment (eileenmcnaughton) [reply tests/phpunit/CRM/Contribute/ActionMapping/ByTypeTest.php:263]: @seamuslee001 yeah - good question - I've figured out it's coming from here 

https://github.com/civicrm/civicrm-core/blob/26219967a075d18e87fbfa84d4ce99e08de89e17/Civi/Token/TokenRow.php#L208-L212

(that function should be moved to Contribute_Tokens IMHO since it's not used from elsewhere & it's got some separation of concern issues)

So I just need to check what buildOptions would return there

Comment (eileenmcnaughton) [reply tests/phpunit/CRM/Contribute/ActionMapping/ByTypeTest.php:263]: Aha -       'get' => "get: all options are returned, even if they are disabled; labels are translated.", - so it looks like it should be label

Comment (seamuslee001) [reply tests/phpunit/CRM/Contribute/ActionMapping/ByTypeTest.php:263]: Maybe worth altering the label in the test just to confirm that perhaps?

Comment (eileenmcnaughton) [reply tests/phpunit/CRM/Contribute/ActionMapping/ByTypeTest.php:263]: hmm - you mean just as a way to test locally to double check my read of the code?

Comment (seamuslee001) [reply tests/phpunit/CRM/Contribute/ActionMapping/ByTypeTest.php:263]: Kind of / lock in that behaviour?

Comment (eileenmcnaughton) [reply tests/phpunit/CRM/Contribute/ActionMapping/ByTypeTest.php:263]: @seamuslee001 I'm confused - because it's behaviour being removed so it can't be locked in?

Comment (seamuslee001) [reply tests/phpunit/CRM/Contribute/ActionMapping/ByTypeTest.php:263]: @eileenmcnaughton this is what I was thinking of https://github.com/civicrm/civicrm-core/commit/9fbcf5754e56234393aed03e9ce361e6a989c345, I would also suggest that in your upgrade script you change reference from :name to :label to match what was actually going on

Comment (eileenmcnaughton) [reply tests/phpunit/CRM/Contribute/ActionMapping/ByTypeTest.php:263]: @seamuslee001 yeah we DO want to remove that token altogether though rather than lock it in. I can confirm locally that it IS label but I've already updated the PR for it to be label.

File: tests/phpunit/CRM/Contribute/ActionMapping/ByTypeTest.php [hunk 3]
```diff
@@ -232,9 +232,9 @@ public function scheduleForSoftCreditor() {
   }
 
   public function useHelloFirstNameStatus() {
-    $this->schedule->subject = 'Hello, {contact.first_name}. @{contribution.status}. (via subject)';
-    $this->schedule->body_html = '<p>Hello, {contact.first_name}. @{contribution.status}. (via body_html)</p>';
-    $this->schedule->body_text = 'Hello, {contact.first_name}. @{contribution.status} (via body_text)';
+    $this->schedule->subject = 'Hello, {contact.first_name}. @{contribution.contribution_status_id:name}. (via subject)';
```
Comment (seamuslee001) [inline tests/phpunit/CRM/Contribute/ActionMapping/ByTypeTest.php:235]: @eileenmcnaughton shouldn't this shift to be label?

File: tests/phpunit/CRM/Upgrade/Incremental/BaseTest.php [hunk 4]
```diff
@@ -64,6 +66,33 @@ public function testMessageTemplateStringReplace(): void {
...
+      'start_action_date' => '2021-01-01',
+      'mapping_id' => 1,
+      'entity_value' => 1,
+      'body_text' => 'blah {contribution.status}',
+      'body_html' => 'blah {contribution.status}',
+      'subject' => 'blah {contribution.status}',
+    ])->execute();
+
+    $upgrader = new CRM_Upgrade_Incremental_MessageTemplates('5.41.0');
+    $upgrader->replaceTokenInActionSchedule('contribution.status', 'contribution.contribution_status_id:name');
```
Comment (seamuslee001) [inline tests/phpunit/CRM/Upgrade/Incremental/BaseTest.php:87]: @eileenmcnaughton shouldn't we change reference from :name to label here as well?

File: tests/phpunit/CRM/Contribute/ActionMapping/ByTypeTest.php (hunk 3 above)
Comment (eileenmcnaughton) [reply tests/phpunit/CRM/Contribute/ActionMapping/ByTypeTest.php:235]: @seamuslee001 yeah - I can do - the follow up patches I have sitting around use a combo of name & label for the various fields

File: tests/phpunit/CRM/Upgrade/Incremental/BaseTest.php (hunk 4 above)
Comment (eileenmcnaughton) [reply tests/phpunit/CRM/Upgrade/Incremental/BaseTest.php:87]: hmm - well in that case it's testing the functionality - we will wind up with a range of substitutions so it's not the data that matters per se

File: tests/phpunit/CRM/Contribute/ActionMapping/ByTypeTest.php (hunk 2 above)
Comment (seamuslee001) [reply tests/phpunit/CRM/Contribute/ActionMapping/ByTypeTest.php:263]: @eileenmcnaughton I guess my point was more that at the present moment we aren't testing that it is the actual label to be returned and also not sure if we are testing anywhere that the translation works as well but I guess it felt like to me that we were swapping apples for oranges (going from testing that the label is outputted from {contribution.status} to using name {contribution.contribution_status_id:name})

File: tests/phpunit/CRM/Upgrade/Incremental/BaseTest.php (hunk 4 above)
Comment (seamuslee001) [reply tests/phpunit/CRM/Upgrade/Incremental/BaseTest.php:87]: I just figure if anyone was looking this is kind of testing the upgrade step that we are adding so shouldn't we want it to match the upgrade step that we are doing?

File: tests/phpunit/CRM/Contribute/ActionMapping/ByTypeTest.php (hunk 3 above)
Comment (eileenmcnaughton) [reply tests/phpunit/CRM/Contribute/ActionMapping/ByTypeTest.php:235]: @seamuslee001 OK - I've added specific tests for label into the PR - I originally considered removing the status token from this line but left it in - however it's just an extra place that token is tested rather than anything meaningful

File: tests/phpunit/CRM/Contribute/ActionMapping/ByTypeTest.php (hunk 2 above)
Comment (eileenmcnaughton) [reply tests/phpunit/CRM/Contribute/ActionMapping/ByTypeTest.php:263]: @seamuslee001 yeah I added label in - the follow up changes put financial_type_id through the same process & that one was testing label but I guess that one isn't an open PR yet since it was pending this one so I just added in label here

File: tests/phpunit/CRM/Upgrade/Incremental/BaseTest.php (hunk 4 above)
Comment (eileenmcnaughton) [reply tests/phpunit/CRM/Upgrade/Incremental/BaseTest.php]: Hmm - I'm not convinced it's better but it's not worse to switch to label so I've done it

**Comment by seamuslee001** [comment]
This works for me added merge on pass

**Comment by seamuslee001** [comment]
Jenkins re test this please
